### ISSUE TITLE
QEP  29: QGIS 3.0

### ISSUE THREAD

**Issue by NathanW2** [qgis/QGIS-Enhancement-Proposals#29]
# 
# QGIS Enhancement 29: QGIS 3.0

**Date** 2015/10/21
**Author** Nyall Dawson
**Contact** nyall dot dawson at gmail dot com
**Last Edited** 2015/10/21
**Status**  Draft
**Version** QGIS 3.0
## 1. Summary

Following the LTR 2.14 release, the subsequent release should be QGIS 3.0, and include:
- an API break
- Qt5 only build support
- Python 3.0 support
## 3. Proposed Changes

3.1 API Break
.............

QGIS API has remained stable throughout the 2.0 series. While it is acknowledged
that this is desirable for plugin and script maintainers, QGIS 3.0 will allow 
API breaks. This is required to clean up the codebase and make deeper changes
and bug fixes which are currently impossible to do without breaking API. It
also allows for removal of old compatibility code (currently QGIS master has over
300 methods marked as deprecated), simplifying future development and 
maintenance efforts.

3.2 Qt 5 Requirement
....................

Currently QGIS supports building on both Qt4 and Qt5 libraries, with the default
build being Qt4. Qt4 has been end-of-lifed by it's developers and is no longer
maintained. This raises significant risks for QGIS if we do not move to Qt5 builds
as quickly as possible. For instance, Qt4 cannot be built on the latest OSX 
release (10.11). Third party patches have been made which partly resolve this
issue, but the side effects are unknown. It is highly likely that a future
OSX release, Windows update or linux packaging change will prevent use of
Qt4 on one of our supported platforms and result in QGIS being unavailable
for that platform.

Additionally, numerous bugs present in the current QGIS release which are
caused by issues in Qt have been resolved in Qt5. Moving to Qt5 is the only way
to fix these issues.

There's also significant performance and feature benefits made possible by
the switch to the newer platform.

3.2.1 Issues with moving to Qt5
...............................

Qt5 has dropped support for the QWebView/QWebPage classes. The replacement
class, QtWebEngine, is immature and lacks several important features which
are currently used by QGIS. The largest impact this will have will be on
the composer "HTML" item. This item is currently rendered using a QWebPage
painting directly on to a QPainter surface, resulting in text rendering
as text objects and vectors. This is NOT currently possible with the 
replacement classes, and it is unclear whether this will be brought back
by a future Qt release. It is also impossible to render QtWebEngine contents
using a transparent background. Unfortunately, there is no workaround for these
limitations and the feature will have to be modified so that HTML content is either:
- rendered as an upscaled (read: pixelated) image
- rendered using the subset of HTML and CSS supported by QTextDocument (see
  http://doc.qt.io/qt-5/richtext-html-subset.html )

3.3 Python 3.0
..............

Work is already underway and a PR to master is inplace to allow QGIS to be built using Python 3.0 and PyQt5.
## 4.0 Implications for 2.14 LTR release

The QGIS 3.0 release will follow the LTR 2.14 release, which will be the final release in the 2.0 series. In order to support organisations who will be stuck using the 2.0 series for the immediate future (eg due to plugin requirements or other script support requirements) only low-risk, small impact changes will be allowed in QGIS 2.14. This is intended to allow us to deliver the most-stable version possible as the final 2.0 series release.

Large changes should be held-off until master is un-frozen for 3.0 development. 
# .# Affected Files

..................

Almost everything!
# . Further Considerations/Improvements

---

Support for plugin developers in the form of guides for porting to QGIS 3.0
will need to be developed.
# . Voting History

---

(required)

**Comment by nyalldawson** [comment]
@NathanW2 I've update this QEP proposal. Basically I've stripped out everything contentious (eg double-length development cycle) and left in two main points:
- after 2.14 we will release 3.0, which is an API break for Qt5/Python 3.0/PyQt5 requirements
- only low-risk/low-impact changes allowed for 2.14

This should allow us to pass this proposal quickly, and then further considerations (eg timing for 3.0 release - double cycle or not?) can be discussed as seperate issues. The important thing first is to determine that the next release after 2.14 will be 3.0...

**Comment by nyalldawson** [comment]
@NathanW2 in other words, this is ready for voting

**Comment by timlinux** [comment]
+1 from me - nice simple and clear roadmap. Just worried about what will happen without web views...many plugins out there probably rely on it too - for example InaSAFE (which I work on) uses them heavily in the UI and in composer.

**Comment by nyalldawson** [comment]
@timlinux yes, it's really unfortunate, and I'd really hoped Qt would have sorted this mess out now that Qt5 has been out a while. Painful, but totally out of our control and something we'll/our users will just have to live with.

**Comment by giohappy** [comment]
@NathanW2 I don't know Qt5 very much but are the current QWebView/QWebPage so different from Qt4's? http://doc.qt.io/qt-5/qwebview.html

**Comment by nyalldawson** [comment]
@giohappy see http://stackoverflow.com/questions/29055475/qwebview-or-qwebengineview - QWebView has also been deprecated and will be removed from Qt as of 5.6 (see https://wiki.qt.io/New_Features_in_Qt_5.6). We could use it temporarily, but can't rely on it being available on all our platforms/Qt versions. The future is QtWebEngine, which (currently) has no support for printing or rendering webpages. It's on their TODO list for 5.7 (see https://trello.com/c/JE5kosmC/72-printing-support), but given:
1. it's been continually pushed back since 5.4 - who knows when it will eventually land, and if it will be usable for our needs
2. we'll have to support Qt5 < 5.7 (or whenever it lands) for the forseeable future anyway, so can't rely on this.

In short - It's a huge mess, and it won't be sorted out by Qt/distros/etc before QGIS 3.0.

**Comment by giohappy** [comment]
Thanks @NathanW2 now it's clear. I see that a lot of improvements have been
made on 3D and OpenGL support: stop doing web dev and let's focus on 3D :D
Il 10/nov/2015 23:15, "Nyall Dawson" <notifications@github.com> ha scritto:

**Comment by 3nids** [comment]
* What would be the developpment / freeze time for 3.0? Same as the others or longer?
* Is there a list of proposed major changes / API break?
* For Qt5, does #2430 will be merged soon (in 2.14) so we can start building/testing with Qt5/Py3 now and make it a requirement for 3.0?

**Comment by nyalldawson** [comment]
>What would be the developpment / freeze time for 3.0? Same as the others or longer?

That's been deliberately left out of scope for this QEP - the important thing to decide first is that we'll go 2.14 -> 3.0, and what that means. Timing can be discussed in a followup QEP, but I want to avoid this getting side tracked in other details.

> Is there a list of proposed major changes / API break?

Same as above... (apart from the mentioned Qt5/Python 3.0 minimum version requirements)

> For Qt5, does #2430 will be merged soon (in 2.14) so we can start building/testing with Qt5/Py3 now and make it a requirement for 3.0?

That's the plan

**Comment by 3nids** [comment]
makes sense to me, thanks for the clarification!

**Comment by NathanW2** [comment]
@timlinux do you think we can make a vote thing for this?

Any comments from anyone else before we go to that step?

**Comment by nyalldawson** [comment]
@NathanW2  This should be rejected, right?

**Comment by timlinux** [comment]
Hi @nyalldawson 

> @NathanW2 This should be rejected, right?

Yes I think so - according to our plan, we will first release 2.16 with wrappers for Qt5 support and then deprecate Qt4 in the subsequent 3.0 release.

**Comment by Juanlu001** [comment]
Apparently this does not reflect the current roadmap anymore:

https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule
